### ISSUE TITLE
Query: QueryFilter with EntityTypeConfiguration are failing to inject current context values

### ISSUE THREAD

**Issue by smitpatel** [aspnet/EntityFrameworkCore#10301]
```C#
using System;
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;
using Microsoft.Extensions.Logging;

namespace EFSampleApp
{
    public class Program
    {
        public static void Main(string[] args)
        {


            using (var db = new MyContext())
            {
                // Recreate database
                db.Database.EnsureDeleted();
                db.Database.EnsureCreated();

                // Seed database


                db.SaveChanges();
            }

            using (var db = new MyContext())
            {
                // Run queries
                db.Blogs.ToList();
                db.Indirection.TenantId = 1;
                db.Blogs.ToList();
                db.Indirection.TenantId = 2;
                db.Blogs.ToList();
                db.Indirection.TenantId = 3;
                db.Blogs.ToList();
            }

            Console.WriteLine("Program finished.");
        }
    }


    public class MyContext : DbContext
    {
        private static ILoggerFactory LoggerFactory => new LoggerFactory().AddConsole(LogLevel.Trace);
        public Indirection Indirection = new Indirection();

        // Declare DBSets
        public DbSet<Blog> Blogs { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            // Select 1 provider
            optionsBuilder
                .UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=_ModelApp;Trusted_Connection=True;Connect Timeout=5;ConnectRetryCount=0")
                //.UseSqlite("filename=_modelApp.db")
                //.UseInMemoryDatabase(databaseName: "_modelApp")
                .EnableSensitiveDataLogging()
                .UseLoggerFactory(LoggerFactory);
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            // Configure model
            new Configuration(this).Configure(modelBuilder.Entity<Blog>());

        }
    }

    public class Configuration : IEntityTypeConfiguration<Blog>
    {
        public Configuration(MyContext context)
        {
            Context = context;
        }

        public MyContext Context { get; }

        public void Configure(EntityTypeBuilder<Blog> builder)
        {
            builder.HasQueryFilter(e => e.Id == Context.Indirection.TenantId);
        }
    }

    public class Indirection
    {
        public int TenantId { get; set; }
    }

    public class Blog
    {
        public int Id { get; set; }
    }
}

```
The generated SQL is same for all cases causing incorrect results in version: 2.0.0
& throws exception as in #10271 

Work-around: Configure filter in `OnModelCreating` directly.

**Comment by smitpatel** [comment]
If we use single level property off context e.g. `Context.TenantId` rather than indirection, then this worked correctly in 2.0.0

**Comment by granthoff1107** [comment]
@ ajcvickers I'm receiving a similar Issue to this even when using the work around through OnModelCreating.   https://github.com/aspnet/EntityFrameworkCore/issues/10709,  is there a temporary fix for this or a work around, or pre-release of 2.1.0 or should I just end up using EntityFramework-Plus?

**Comment by ajcvickers** [comment]
@granthoff1107 This issue isn't fixed yet--hence it is still open. #10709 looks somewhat different although I agree that it could be the same root cause. We'll look at it in triage.

**Comment by mhosman** [comment]
Same error here... InvalidCastException: Unable to cast object of type 'System.Int32' to type 'System.Linq.Expressions.LambdaExpression'.

**Comment by smitpatel** [comment]
## Design meeting notes

This issue will solve following:

- We will try to get current context type insider ParameterExtractingEV so that we can capture instances of only that context type in filter. Any other context type member access will be inlined (or throw exception if it is too difficult)

- Patterns of configuration which would capture context correctly and inject current instance values
  - Defining filter in OnModelCreating
  - Defining filter in EntityTypeConfiguration by passing context through constuctor
  - Defining filter using method (inside/outside DbContext or extension method) where context is passed as parameter.
  - Any of above where context is wrapped inside another object type and that type is being passed around.

- Apart from above we will parametrize any kind of call on DbContext i.e. property/field access, method call, going through multiple levels.

Future goals:
- Perhaps a HasQueryFilter overload which takes lambda with derived DbContextType parameter, explicitly indicating which DbContext instance has to be captured and parametrized. Filed #11040

**Comment by smitpatel** [comment]
The fix for this is available in nightly build 2.1.0-preview2-30200. If people can give it a try and bring anymore missing cases to our attention, it would help us in improving it further.

**Comment by AliCharper** [comment]
> The fix for this is available in nightly build 2.1.0-preview2-30200. If people can give it a try and bring anymore missing cases to our attention, it would help us in improving it further.

We are using EF 2.2 and the issue is still there. So how ?.

**Comment by ajcvickers** [comment]
@AliCharper Please file a new issue including a small, runnable project/solution or complete code listing that demonstrates the behavior you are seeing.
